File: Dbus/interfaces/ruby/CDApplet.rb
#!/usr/bin/ruby

# This is a part of the external Ruby Battery applet for Cairo-Dock
#
# Author: Eduardo Mucelli Rezende Oliveira
# E-mail: edumucelli@gmail.com or eduardom@dcc.ufmg.br
#
# This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.

# This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.

# This applet monitors the battery through acpi module. It is possbile to show a dialog message
#    containing the baterry status, charge, and temperature by left-clicking on the icon.
#    Also, through the configuration panel it is possible to set the icon's label, 
#	 show the % of carge as quick info, and activate an alert message to be shown when the charge is critically low.

require 'rubygems'
require 'dbus'
require 'parseconfig'

class CDApplet
	attr_accessor :conf_file, :applet_name, :icon, :sub_icons, :config, :bus
	
	def initialize
		self.applet_name = File.basename(Dir.getwd)
		self.conf_file = File.expand_path("~/.config/cairo-dock/current_theme/plug-ins/#{self.applet_name}/#{self.applet_name}.conf")
		self.config = {}
		self.bus = nil
		self.icon = nil
		self.sub_icons = nil
		
		self._get_config()
		self._connect_to_dock()
	end
	
	def run
		self.start()
		loop = DBus::Main.new
		loop << self.bus
		loop.run
		puts ">>> applet '#{self.applet_name}' terminated."
		exit
	end
	
	##################################
	### callbacks on the main icon ###
	##################################
	
	def on_click iState
		### action on click
	end
	
	def on_middle_click
		puts ">>> on_middle_click"
		### action on middle-click
	end
	
	def on_build_menu
		### build our menu
	end
	
	def on_menu_select(iNumEntry)
		### action on selecting an entry of our menu """
	end
	
	def on_scroll(bScrollUp)
		### action on scroll
	end

	def on_drop_data(cReceivedData)
		### action on dropping something on our applet
	end
	
	def on_answer(answer)
		### action on answering a dialog
	end

	def on_answer_dialog(button, answer)
		### action on answering a dialog
	end

	def on_shortkey(cKey)
		### action on pressing one of the shortkeys we bound beforehand
	end
	
	def on_change_focus(bIsActive)
		### action when the window controlled by the applet takes or looses the focus
	end
	
	##################################
	### callbacks on the sub-icons ###
	##################################
	def on_click_sub_icon(iState, cIconID)
		### action on click on one of our sub-icons
	end
	
	###############################
	### callbacks on the applet ###
	###############################
	
	def start
		### action when the applet is started
	end
	
	def stop
		### action when the applet is terminated
	end
	
	def _on_stop
		self.stop()
		exit
	end
	
	def reload
		### called when our applet is reloaded (config parameters have changed)
	end
	
	def _on_reload bConfigHasChanged
		if bConfigHasChanged
			self._get_config()
			self.reload()
		end
	end
	
	def _get_config
		keyfile = ParseConfig.new(self.conf_file)
		self.get_config(keyfile)
	end
	
	def get_config keyfile
		### get our parameters from the key-file
	end
	
	def _connect_to_dock
		# get the icon object on the bus
		self.bus = DBus::SessionBus.instance
		applet_path = "/org/cairodock/CairoDock/#{self.applet_name}" # path where our object is stored on the bus
		applet_service = bus.service("org.cairodock.CairoDock")
		begin
			applet_object = applet_service.object(applet_path)
			applet_object.introspect
			applet_object.default_iface = 'org.cairodock.CairoDock.applet'
		rescue
			puts ">>> object '#{self.applet_name}' can't be found on the bus, exit.\nMake sure that the 'Dbus' plug-in is activated in Cairo-Dock"
			exit
		end
		self.icon = applet_object
		
		# get the sub-icons object on the bus
		applet_sub_icons_object = applet_service.object("#{applet_path}/sub_icons")
		applet_sub_icons_object.introspect
		applet_sub_icons_object.default_iface = 'org.cairodock.CairoDock.subapplet'
		self.sub_icons = applet_sub_icons_object
		
		# now connect to the signals
		self.icon.on_signal("on_click") do |iState|
			self.on_click iState
		end
		self.icon.on_signal("on_middle_click") do
			self.on_middle_click
		end
		self.icon.on_signal("on_build_menu") do
			self.on_build_menu
		end
		self.icon.on_signal("on_menu_select") do |iNumEntry|
			self.on_menu_select iNumEntry
		end
		self.icon.on_signal("on_scroll") do |bScrollUp|
			self.on_scroll bScrollUp
		end
		self.icon.on_signal("on_drop_data") do |cReceivedData|
			self.on_drop_data cReceivedData
		end
		self.icon.on_signal("on_answer") do |answer|
			self.on_answer answer
		end
		self.icon.on_signal("on_answer_dialog") do |button, answer|
			self.on_answer_dialog button, answer
		end
		self.icon.on_signal("on_shortkey") do |cKey|
			self.on_shortkey cKey
		end
		self.icon.on_signal("on_change_focus") do |bIsActive|
			self.on_change_focus bIsActive
		end
		
		self.icon.on_signal("on_stop_module") do
			self._on_stop
		end
		
		self.icon.on_signal("on_reload_module") do |bConfigHasChanged|
			self._on_reload bConfigHasChanged
		end
		
		self.sub_icons.on_signal("on_click_sub_icon") do |iState, sub_icon_id|
			self.on_click_sub_icon iState sub_icon_id
		end
	end
end

File: Dbus/interfaces/python/CDApplet.py
# This is a part of the external demo applet for Cairo-Dock
# Copyright : (C) 2010-2011 by Fabounet
# E-mail : fabounet@glx-dock.org
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
# http://www.gnu.org/licenses/licenses.html#GPL

# Base class for Cairo-Dock's applets.
# Make your own class derive from a CDApplet, and override the functions you need (the ones which don't start with an underscore).

####################
### dependancies ###
####################
import sys
import os.path
import ConfigParser
import gobject
import glib
import dbus
from dbus.mainloop.glib import DBusGMainLoop

DBusGMainLoop(set_as_default=True)

####################
### Applet class ###
####################
class CDApplet:
	
	BOTTOM = 0
	TOP    = 1
	RIGHT  = 2
	LEFT   = 3
	DOCK    = 0
	DESKLET = 1
	UPPER_LEFT  = 0
	LOWER_RIGHT = 1
	LOWER_LEFT  = 2
	UPPER_RIGHT = 3
	MIDDLE      = 4

	
	def __init__(self,applet_name=None):
		""" initialize the applet. Must be called by any class that inheritates from it.
		It defines the following:
		 - icon : our main icon
		 - sub_icons: our sub-icons
		 - config : a dictionnary where our configuration parameters are stored
		 - applet_name : name of our applet (the same as the folder where files are stored)
		 - conf_file : path to our config file (you will rarely need it)
		 """
		self.icon = None
		self.sub_icons = None
		self.config = {}
		if applet_name is None:
			self.applet_name = os.path.basename(os.path.abspath("."))  # the name of the applet must the same as the folder.
		else:
			self.applet_name = applet_name
		self.conf_file = os.path.expanduser("~/.config/cairo-dock/current_theme/plug-ins/"+self.applet_name+"/"+self.applet_name+".conf")  # path to the conf file of our applet.
		
		self._get_config()
		self._connect_to_dock()
	
	def run(self):
		""" start the applet and enter the main loop; we never get out of this function """
		self.begin()
		self.loop = gobject.MainLoop()
		self.loop.run()
		print ">>> applet '"+self.applet_name+"' terminated."
		sys.exit(0)
	
	##################################
	### callbacks on the main icon ###
	##################################
	def on_click(self,iState):
		""" action on click """
		pass
	
	def on_middle_click(self):
		""" action on middle-click """
		pass
	
	def on_build_menu(self):
		""" build our menu """
		pass
		
	def on_menu_select(self,iNumEntry):
		""" action on selecting an entry of our menu """
		pass
	
	def on_scroll(self,bScrollUp):
		""" action on scroll """
		pass
	
	def on_drop_data(self,cReceivedData):
		""" action on dropping something on our applet """
	
	def on_answer(self,answer):
		""" action on answering a dialog """
		pass
	
	def on_answer_dialog(self, button, answer):
		""" action on answering a dialog """
		pass
	
	def on_shortkey(self,cKey):
		""" action on pressing one of the shortkeys we bound beforehand """
		pass
	
	def on_change_focus(self,bIsActive):
		""" action when the window controlled by the applet takes or looses the focus """
		pass
	
	##################################
	### callbacks on the sub-icons ###
	##################################
	def on_click_sub_icon(self, iState, cIconID):
		""" action on click on one of our sub-icons"""
		pass
	
	###############################
	### callbacks on the applet ###
	###############################
	def begin(self):
		""" action when the applet is started """
		pass
	
	def end(self):
		""" action when the applet is terminated """
		pass
	
	def _on_stop(self):
		self.end()
		self.loop.quit()
	
	def reload(self):
		""" called when our applet is reloaded (config parameters have changed) """
		pass
	
	def _on_reload(self,bConfigHasChanged):
		if bConfigHasChanged:
			self._get_config()
			self.reload()
	
	def get_config(self,keyfile):
		""" get our parameters from the key-file """
		pass
	
	def _get_config(self):
		keyfile = ConfigParser.RawConfigParser()
		keyfile.read(self.conf_file)
		self.get_config(keyfile)
	
	def _connect_to_dock(self):
		# get our applet on the bus.
		print "self.applet_name:"+self.applet_name
		applet_path = "/org/cairodock/CairoDock/"+self.applet_name  # path where our object is stored on the bus.
		bus = dbus.SessionBus()
		try:
			applet_object = bus.get_object("org.cairodock.CairoDock", applet_path)
		except:
			print ">>> object '"+self.applet_name+"' can't be found on the bus, exit.\nMake sure that the 'Dbus' plug-in is activated in Cairo-Dock"
			sys.exit(2)
		self.icon = dbus.Interface(applet_object, "org.cairodock.CairoDock.applet")  # this object represents our icon inside the dock or a desklet.
		sub_icons_object = bus.get_object("org.cairodock.CairoDock", applet_path+"/sub_icons")
		self.sub_icons = dbus.Interface(sub_icons_object, "org.cairodock.CairoDock.subapplet")  # this object represents the list of icons contained in our sub-dock, or in our desklet. We'll add them one by one later, giving them a unique ID, which will be used to identify each of them.
		# connect to signals.
		self.icon.connect_to_signal("on_click", self.on_click)  # when the user left-clicks on our icon.
		self.icon.connect_to_signal("on_middle_click", self.on_middle_click)  # when the user middle-clicks on our icon.
		self.icon.connect_to_signal("on_build_menu", self.on_build_menu)  # when the user right-clicks on our applet (which builds the menu)
		self.icon.connect_to_signal("on_menu_select", self.on_menu_select)  # when the user selects an entry of this menu.
		self.icon.connect_to_signal("on_scroll", self.on_scroll)  # when the user scroll up or down on our icon.
		self.icon.connect_to_signal("on_drop_data", self.on_drop_data)  # when the user drops something on our icon.
		self.icon.connect_to_signal("on_answer", self.on_answer)  # when the user answer a question.
		self.icon.connect_to_signal("on_answer_dialog", self.on_answer_dialog)  # when the user answer a question.
		self.icon.connect_to_signal("on_shortkey", self.on_shortkey)  # when the user press the shortkey.
		self.icon.connect_to_signal("on_change_focus", self.on_change_focus)  # when the window's focus changes.
		self.icon.connect_to_signal("on_stop_module", self._on_stop)  # when the user deactivate our applet (or the DBus plug-in, or when the Cairo-Dock is stopped).
		self.icon.connect_to_signal("on_reload_module", self._on_reload)  # when the user changes something in our config, or when the desklet is resized (with no change in the config).
		self.sub_icons.connect_to_signal("on_click_sub_icon", self.on_click_sub_icon)  # when the user left-clicks on a sub-icon.